Expose various function in `tags` publicly

brettcannon opened the pull request:
Closes #187 and should open the way for pip to start using packaging.tags.

xavfernandez commented on packaging/tags.py:
```diff
@@ -433,71 +528,62 @@ def _linux_platforms(is_32bit=_32_BIT_INTERPRETER):
+    warn = _warn_parameter("sys_tags", kwargs)
 
+    interpreter_name = _interpreter_name()
     if interpreter_name == "cp":
```
Maybe `"cp"` & `"pp"` along with the other prefixes defined in the PEP 425 could be define as constant in the module ?

brettcannon commented:
This should also close #184

brettcannon commented:
I will also note that `_generic_abi()` could probably be changed to return a sequence and leave out `"none"`.

brettcannon commented on packaging/tags.py:
```diff
@@ -289,11 +379,16 @@ def _mac_binary_formats(version, cpu_arch):
+    """
+    Yields the platform tags for a macOS system.
+
+    The *version* parameter is a two-item tuple specifying the macOS version to
```
You tell me. 😉 In the stdlib we use `*` and not backticks. And I checked the docs for  'packaging' and the closest I can see is using double-backtick, so we're both right/wrong? 🤷‍♂

pradyunsg replied:
lol

Let's do double backticks, for the sake of ReST?

brettcannon replied:
Actually I was talking about the docs (review comments got interleaved in an odd way). If you're going to care for docstrings then I would much rather use  single backticks.

brettcannon commented on docs/tags.rst:
```diff
@@ -92,11 +93,102 @@ Reference
+                            PyPy interpreter.
+    :param Iterable abis: Iterable of compatible ABIs. Defaults to the ABIs
+                          compatible with the current system.
+    :param Iterable platforms: Iterable of compatible platforms. Defaults to the
```
It only works if there's no spaces in the name. So for instance this fails for `Tuple[int, int]`, and so I didn't bother adding type parameters when it will only work in some instances.

pradyunsg replied:
Ah well, that makes sense.

I think we can visit how we're documenting these in a follow up issue/PR.

brettcannon commented on docs/tags.rst:
```diff
@@ -92,11 +93,102 @@ Reference
+
+    :param str interpreter: The name of the interpreter. Defaults the current
+                            PyPy interpreter.
+    :param Iterable abis: Iterable of compatible ABIs. Defaults to the ABIs
```
See above.

brettcannon commented on packaging/tags.py:
```diff
@@ -162,20 +172,55 @@ def _cpython_abis(py_version, warn=False):
-def _cpython_tags(py_version, interpreter, abis, platforms):
-    # type: (PythonVersion, str, Iterable[str], Iterable[str]) -> Iterator[Tag]
+def cpython_tags(
+    python_version=sys.version_info[:2],  # type: PythonVersion
```
Sure, but can I ask why? This avoids that `if` check and I have gotten pushback previously in this repo for having too many explicit conditional branches. Plus tuples are immutable so there's also no worry about accidentally mutating the constant in-place. Is it for consistency reasons?

brettcannon replied:
And does that mean you prefer an `if ... is None` check instead to tightly check for `None` versus any false value?

chrahunt replied:
I :+1:d because it makes the fact that it defaults to `sys.version_info` testable vs the current construction.

pradyunsg replied:
Yea, as @chrahunt points out, my main thought was testability of the fact that we're using sys.version_info for this default (possibly via mocking out the dependency to confirm that).

That said, I don't feel strongly about this since that's the only reason it's useful in that one case and IIUC, the benefit of doing so isn't much either.

chrahunt replied:
I'm trying to finish up the integration against this branch now. Given that context, I think it is easier to use if defaulted to `None` here. For all other parameters we can do something like:

```
abis = None
if abi_from_user:
    abis = [abi_from_user]

cpython_tags(
    ...,
    abis=abis,
    ...
)
```

but for this one we'd need to either:

1. explicitly default to `python_version = sys.version_info[:2]` outside the function and have a test that we defaulted correctly, or
2. create a dictionary `args = {}` and conditionally do `args["python_version"] = python_version_from_user` and invoke this function like `cpython_tags(**args)`, which I don't think is friendly with type checkers without additional work to define the type of `args`

I would make the same argument for having `python_version=None` in `compatible_tags` too.

brettcannon commented on docs/tags.rst:
```diff
@@ -92,11 +93,102 @@ Reference
+    determined statically. For Windows this information is entirely static and
+    thus does not require calculating older version support. For Linux, code
+    must be run on the system itself to determine its compatibility level and
+    thus cannot be calculated statically.
```
I went another way and just made it a note as it actually isn't important to how the function operates (it's just to keep people from asking "but what about Linux?").

brettcannon commented on packaging/tags.py:
```diff
@@ -433,71 +528,62 @@ def _linux_platforms(is_32bit=_32_BIT_INTERPRETER):
+    yield _normalize_string(distutils.util.get_platform())
+
+
+def _platforms():
```
I think that's too verbose, so I'll meet you halfway with `_platform_tags()`. 😄

brettcannon commented on docs/tags.rst:
```diff
@@ -92,11 +93,102 @@ Reference
     :param bool warn: Whether warnings should be logged. Defaults to ``False``.
 
 
+.. function:: mac_platforms(version=None, arch=None)
```
If we go that route then I would want to have parameter annotations which then makes me want to drop the whole `:param` thing as that makes them fairly redundant.

pradyunsg replied:
Cool, let's tackle that in a follow up then. :)

brettcannon commented on docs/tags.rst:
```diff
@@ -92,11 +93,102 @@ Reference
+
+.. function:: generic_tags(interpreter=None, abis=None, platforms=None, *, warn=False)
+
+    Yields the tags for an interpreter in a non-specialized fashion.
```
It's anything not CPython. 😄 I'll tweak it.

brettcannon commented on packaging/tags.py:
```diff
@@ -105,8 +105,24 @@ def parse_tag(tag):
     return frozenset(tags)
 
 
+def _warn_parameter(func_name, kwargs):
```
Once again, met you halfway. :)

brettcannon commented on packaging/tags.py:
```diff
@@ -200,53 +245,98 @@ def _generic_abi():
-    all following versions up to 'end'.
+    all previous versions of that major version.
     """
     yield "py{major}{minor}".format(major=py_version[0], minor=py_version[1])
```
This can also be done with `filter()`.

The key reason I'm hesitant is it will make `generic_tags()` different from `cpython_tags()` in terms of supporting a major-only `python_version` argument. And so documenting that unique distinction will have to be made very clear and some unique parameter name will be necessary so people don't assume there's a naming convention.

chrahunt replied:
So callers should provide a dummy value for minor version in both cases (`compatible_tags()` and `cpython_tags()`)?

brettcannon replied:
Fair point about the potential problem of future semantics. Unfortunately I still don't have a good answer for my other concerns either. 😞

chrahunt replied:
Is the problem that we can't come up with reasonable behavior in major and (major, minor) cases for both functions? I assume it would be whatever is generally applicable regardless of minor version.

brettcannon replied:
Correct. There is no concept of major-only CPython compatibility and so this doesn't really apply to `cpython_tags()` while it could make sense to `compatibility_tags()`. So that would make both having a parameter named `python_version` that actually have different type requirements a bit confusing. So we would need to come up with some other parameter name for one of them to make this work. Maybe `python_feature_version` for `cpython_tags()`?

chrahunt replied:
Wouldn't `cp3-abi3-*` make sense? Maybe that could be the return value for `cpython_tags()` if given a single-part version (which `== 3`), nothing otherwise, and we can avoid the need to decide on a reasonable-sounding parameter name. :)

brettcannon replied:
@chrahunt I'll think about it. I still think `cp3` is worthless, but if it will resolve this unfortunately design issue then I might be willing to just live with it under a pragmatic argument.

brettcannon commented:
OK, I've addressed all the comments I could without further feedback (left those comments unresolved). The diff is also now massive since I did what @pradyunsg and reorganized all of the tests.

chrahunt commented on packaging/tags.py:
```diff
@@ -433,71 +494,59 @@ def _linux_platforms(is_32bit=_32_BIT_INTERPRETER):
+        return _generic_platforms()
 
 
 def _interpreter_name():
```
Can we expose this one as well? Any implementation that can (optionally) take a user-provided implementation name will need to fall back to this in order to invoke the correct one of (`cpython_tags()`, `generic_tags()`) and `compatibie_tags()`

brettcannon replied:
I was somewhat waiting for someone to ask that. 😄 Yes, I think exposing this is harmless.

brettcannon commented:
OK, I think I have addressed everything at this point. Let me know if you need anything for this PR before I merge it.

chrahunt commented on packaging/tags.py:
```diff
@@ -162,91 +183,156 @@ def _cpython_abis(py_version, warn=False):
+    # type: (bool) -> str
+    version = _get_config_var("py_version_nodot", warn)
+    if not version:
+        version = "".join(map(str, sys.version_info[:2]))
```
Can we expose this as `interpreter_version_nodot()`? It would ease the construction of `interpreter` for `generic_tags` and `compatible_tags`, since we take implementation and version separately and fill in one or the other if not provided.

brettcannon replied:
We can expose it as `interpreter_version()` (the `nodot` is an implementation detail).

brettcannon commented:
@chrahunt PTAL

chrahunt commented:
LGTM!

brettcannon commented:
Thanks everyone for working on this! I know it was a slog to get through.

pradyunsg commented:
Hurrah! Many thanks for all the work here folks! ^>^

xavfernandez commented:
A new packaging release (20.0 ?) would be lovely to unlock https://github.com/pypa/pip/pull/7354 :pray:

di commented:
@xavfernandez I think we need to merge https://github.com/pypa/packaging/pull/221 and https://github.com/pypa/packaging/pull/234 first.

chrahunt commented:
Only #221 is required to prevent breaking something that currently works in pip. #234 is a change to existing behavior that could be evaluated independently in its own time if needed.

pradyunsg commented:
@chrahunt In that case, I think we can go ahead and make a release.

@pypa/packaging-committers Any concerns with making a release? If not, let's go ahead and make one. :)

pradyunsg commented:
Whoops -- I just saw #252 (which is now crosslinked, so yay?)

McSinyx commented on docs/tags.rst:
```diff
@@ -92,11 +94,108 @@ Reference
+    :param Iterable platforms: Iterable of compatible platforms. Defaults to the
+                               platforms compatible with the current system.
+
+.. function:: cpython_tags(python_version=None, abis=None, platforms=None, *, warn=False)
+
+    Yields the tags for the CPython interpreter.
+
+    The specific tags generated are:
+
+    - ``cp<python_version>-<abi>-<platform>``
+    - ``cp<python_version>-abi3-<platform>``
+    - ``cp<python_version>-none-<platform>``
+    - ``cp<older version>-abi3-<platform>`` where "older version" is all older
+      minor versions down to Python 3.2 (when ``abi3`` was introduced)
+
```
Sorry for the necro, but why is `cp<older version>` only available for `abi3` and not `none`?

pradyunsg replied:
Please file a new issue for discussing this.

McSinyx replied:
I found GH-311, I think I'll continue the discussion there.
